protocol.md

```diff
@@ -0,0 +1,338 @@
+	- [Content types](#content-types)
+	- [Message types](#message-types)
+	- [Clock vs Timestamp And Message Ordering](#clock-vs-timestamp-and-message-ordering)
+	- [Quoting](#quoting)
+- [Whisper adapter](#whisper-adapter)
+	- [Whisper node configuration](#whisper-node-configuration)
+	- [Keys management](#keys-management)
+	- [Encryption algorithms](#encryption-algorithms)
+	- [Topic](#topic)
+	- [Message encryption](#message-encryption)
+- [Perfect Forward Secrecy (PFS)](#perfect-forward-secrecy-pfs)
+- [Device syncing](#device-syncing)
+- [One-to-one messages](#one-to-one-messages)
+	- [Sending](#sending)
+		- [Sending using PFS](#sending-using-pfs)
+	- [Receiving](#receiving)
+- [Public messages](#public-messages)
+	- [Sending](#sending-1)
+	- [Receiving](#receiving-1)
+- [Group messages](#group-messages)
+- [Offline messages](#offline-messages)
+	- [Anonymity concerns](#anonymity-concerns)
+- [Whisper V6 extensions (or Status Whisper Node)](#whisper-v6-extensions-or-status-whisper-node)
+	- [New RPC methods](#new-rpc-methods)
+
+# Abstract
+
+Ethereum empowers users and developers to interact with totally new kind of applications called Dapps (Decentralized applications). These application allows to interact with the blockchain on a completely new level which is not only about exchanging values but also executing arbitrary logic. This logic can form very sophisticated programs like DAOs (Decentralized autonomous organizations). The missing part here is how users of Dapps can communicate securely and in a decentralized way with each other. Communication is an essential part of any activity. In this document, we specify a secure and decentralized messaging protocol that is capable of running on the Ethereum network.
+
+# Introduction
+
+TBD
+
+# Terminology
+
+* *Client*: a Whisper node implementing the protocol
+* *Whisper node*: an Ethereum node with Whisper V6 enabled (in the case of geth, it's `--shh` option)
+* *Status Whisper node*: an Ethereum node with Whisper V6 enabled and additional Whisper extensions described in [Whisper V6 extensions (or Status Whisper Node)](#whisper-v6-extensions-or-status-whisper-node)
+* *MailServer*: an Ethereum node with Whisper V6 enabled and a mail server registered capable of storing and providing offline messages
+* *Message*: decrypted Whisper message
+* *Envelope*: encrypted message with some metadata like topic and TTL sent between Whisper nodes; a symmetric or asymmetric key is needed to decrypt it and read the payload
+* *Offline message*: an expired envelope stored by a Whisper node permanently
+
+# Basic Assumption
+
+This protocol assumes the following:
+1. There is an Ethereum node that is capable of discovering peers and implements Whisper V6 service.
+2. Participants of a given Whisper network in order to communicate with each other accept messages with lowered PoW value.
+3. Time is synced between all nodes participating in the given network (this is intrinsic Whisper requirement).
```

Comment:
This seems false, what does this mean more precisely?

Reply:
It isn't false, Whisper uses timestamps and PoW to calculate TTLs, so if you timestamp is either too far away in the past or in the future it will get discarded, and moreover, your node will get dropped.

Reply:
That's why we even introduced NTP sync in status-go, but it doesn't work immediately, it takes some time to get the responses, and if the time of your local device is off, say by 2 minutes, you won't be able to participate in the network.

Reply:
NTP != time sync (impossible unless you have atomic time and bounded by some absolute term to XX probability), so we can phrase it more accurately in terms of dependency on wall clock time.

Reply:
okay, so something like "difference between local clocks of nodes is within 30 seconds?"

Reply:
I guess it should be phrased more accurately, it might give the impression that time is *actively* synced among participants.

`Two peers in a whisper network will only be able to communicate if their relative clock drift is within 20s, otherwise messages will be discarded`

An in the implementation section we can mention NTP, although that's not part of the protocol, and any mean (or none) to get your clock within those 20s will do.

Reply:
they aren't just discarded, you (as a peer) are dropped too, as soon as you send an outdated message (or a message from the future).

Reply:
20 seconds is also purely geth's Whisper implementation thing. I will do something like this:
```
3. Time MUST be synced between all nodes participating in the given network (this is intrinsic requirement of the Whisper specification as well). A clock drift between two peers larger than 20 seconds MAY result in discarding incoming messages.
```

Reply:
`Time MUST be synced between all nodes participating` is too absolute I feel and gives you the impression that there's some syncing mechanism provided, I would only specify that `A clock drift between two peers larger than 20 seconds MAY result in discarding incoming messages` and that synchronization of clocks is left to each node and outside the scope of the protocol.

Reply:
That's a fair point. I will change that.